### PR TITLE
Merge duplicate requests

### PR DESCRIPTION
**Pull request by manjanac**
This is my attempt at addressing duplicate requests for model download (ex: user spams a "get download" button on a slow network). Open to suggestions on handling this better :)

### CODE REVIEW

File: FirebaseMLModelDownloader/Sources/ModelDownloader.swift [hunk 1]
```diff
@@ -436,6 +431,33 @@ extension ModelDownloader {
                 self.mainQueueHandler(completion(.failure(error)))
               }
             }
+            /// Stop keeping track of current download task.
+            self.currentDownloadTask.removeValue(forKey: modelName)
+          }
+          /// Merge duplicate requests if there is already a download in progress for the same model.
+          if let downloadTask = self.currentDownloadTask[modelName],
```
Comment (maksymmalyhin) [inline FirebaseMLModelDownloader/Sources/ModelDownloader.swift:438]: Mutating the dictionary is not a thread-safe operation. As the method may be called from any thread by the client we need to ensure the thread-safety e.g. by calling it from a serial dispatch queue.

File: FirebaseMLModelDownloader/Sources/ModelDownloadTask.swift [hunk 2]
```diff
@@ -68,28 +76,45 @@ extension ModelDownloadTask {
...
   }
 
-  func download(progressHandler: ProgressHandler?, completion: @escaping Completion) {
+  /// Check if downloading is not complete for merging requests.
+  func canMergeRequests() -> Bool {
+    return downloadStatus != .complete
+  }
+
+  /// Merge duplicate requests.
+  func merge(newProgressHandler: ProgressHandler? = nil, newCompletion: @escaping Completion) {
```
Comment (maksymmalyhin) [inline FirebaseMLModelDownloader/Sources/ModelDownloadTask.swift:85]: Please add tests for it.

File: FirebaseMLModelDownloader/Sources/ModelDownloader.swift [hunk 3]
```diff
@@ -436,6 +431,33 @@ extension ModelDownloader {
...
+            downloadTask.canMergeRequests() {
+            downloadTask.merge(
+              newProgressHandler: taskProgressHandler,
+              newCompletion: taskCompletion
+            )
+            DeviceLogger.logEvent(level: .debug,
+                                  message: ModelDownloader.DebugDescription.mergingRequests,
+                                  messageCode: .mergeRequests)
+            downloadTask.resume()
+          } else {
```
Comment (maksymmalyhin) [inline FirebaseMLModelDownloader/Sources/ModelDownloader.swift:448]: Could you please add tests for the merge logic.

**Review by maksymmalyhin — COMMENTED** [review]
Mostly LGTM but a couple thing to address.

File: FirebaseMLModelDownloader/Sources/ModelDownloader.swift (hunk 1 above)
Comment (manjanac) [reply FirebaseMLModelDownloader/Sources/ModelDownloader.swift]: Good catch, totally missed this. Please let me know if I'm doing this right.

File: FirebaseMLModelDownloader/Tests/Unit/ModelDownloaderUnitTests.swift [hunk 4]
```diff
@@ -58,10 +58,11 @@ final class ModelDownloaderUnitTests: XCTestCase {
     options.apiKey = MockOptions.apiKey
     options.projectID = MockOptions.projectID
     FirebaseApp.configure(options: options)
-    FirebaseConfiguration.shared.setLoggerLevel(.debug)
+    // FirebaseConfiguration.shared.setLoggerLevel(.debug)
```
Comment (maksymmalyhin) [inline FirebaseMLModelDownloader/Tests/Unit/ModelDownloaderUnitTests.swift:61]: nit: Still needs to be commented?

**Review by maksymmalyhin — APPROVED** [review]
LGTM, thank you for the updates!

**Pull request merged**